File: LeetCode/cpp/219_contains_duplicate_II.cc
/*
Difficulty: Easy
*/

#include <cmath>
#include <unordered_map>
#include <vector>

class Solution {
 public:
  bool containsNearbyDuplicate(const std::vector<int>& nums, const int k) {
    // num -> index
    std::unordered_map<int, int> seenIndices;

    for (int i = 0; i < nums.size(); ++i) {
      const int num = nums[i];
      const auto it = seenIndices.find(num);

      if (it != seenIndices.cend() && std::abs(i - it->second) <= k) {
        return true;
      } else {
        seenIndices[num] = i;
      }
    }

    return false;
  }
};

File: LeetCode/rust/219_contains_duplicate_II.rs
/*
Difficulty: Easy
*/

use std::collections::HashMap;

impl Solution {
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        if k < 0 {
            return false;
        }

        let mut seen_indices = HashMap::<i32, i32>::new();
        for (i, num) in nums.into_iter().enumerate() {
            let i = i as i32;
            if let Some(&matched_idx) = seen_indices.get(&num) {
                if (i - matched_idx).abs() <= k {
                    return true;
                }
            }
            seen_indices.insert(num, i);
        }

        false
    }
}
